gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManagementStateStore.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.orchestration;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.gobblin.exception.QuotaExceededException;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+
+
+/**
+ * An interface to provide abstractions for managing {@link Dag} and {@link org.apache.gobblin.service.modules.flowgraph.Dag.DagNode} states
+ * and allows add/delete and other functions.
+ */
+public interface DagManagementStateStore {
+  Dag<JobExecutionPlan> getDag(String dagId) throws IOException;
+  Dag<JobExecutionPlan> getFailedDag(String dagId) throws IOException;
+  void addDag(String dagId, Dag<JobExecutionPlan> dag);
+  boolean containsDag(String dagId);
+  Dag.DagNode<JobExecutionPlan> getDagNode(String dagNodeId);
+  Dag<JobExecutionPlan> getParentDag(Dag.DagNode<JobExecutionPlan> dagNode);
+  List<Dag.DagNode<JobExecutionPlan>> getDagNodes(String dagId) throws IOException;
+  List<Dag.DagNode<JobExecutionPlan>> getAllDagNodes() throws IOException;
+  boolean addCleanUpDag(String dagId);
+  void addDagNodeState(String dagId, Dag.DagNode<JobExecutionPlan> dagNode);
+  void deleteDagNodeState(String dagId, Dag.DagNode<JobExecutionPlan> dagNode);
+
+  /**
+   * Persist the {@link Dag} to the backing store.
+   * This is not an actual checkpoint but more like a Write-ahead log, where uncommitted job will be persisted
+   * and be picked up again when leader transition happens.
+   * @param dag The dag submitted to {@link DagManager}
+   */
+  void writeCheckpoint(Dag<JobExecutionPlan> dag) throws IOException;
+  void writeFailedDagCheckpoint(Dag<JobExecutionPlan> dag) throws IOException;
+
+  /**
+   * Delete the {@link Dag} from the backing store, typically upon completion of execution.
+   * @param dag The dag completed/cancelled from execution on {@link org.apache.gobblin.runtime.api.SpecExecutor}.
+   */
+  void cleanUp(Dag<JobExecutionPlan> dag) throws IOException;
+  void cleanUpFailedDag(Dag<JobExecutionPlan> dag) throws IOException;
```

Comment:
if these two are merely a convenience method for the two that follow, please provide a default impl (to derive the dagId from the `dag`)